### PR TITLE
Updates to reflect best practices for PR review

### PR DESCRIPTION
**Pull request by trmartin4**
## Objective

With the addition of `CODEOWNERS` and the maturation of the component teams, I felt it would be good to clarify the expectations around PR authors and reviewers.

These updates include some changes to the contract between PR author and reviewers, namely with the expectation that PR reviews should take place with 24 hours and that Github notifications should be the primary source of review requests.  These have come from a running list that I've been compiling with pain points in our current process, gathered from retros and from other meetings.

I'd welcome any feedback and clarifications on this, as I'd like this PR to serve as a prompt for discussion prior to implementing any of these process changes.  If agreed upon, I plan to update the teams in Slack as well to clarify and answer any questions.

### CODE REVIEW

File: docs/contributing/pull-requests/index.md
Comment (eliykat) [inline docs/contributing/pull-requests/index.md:147]: Has the team agreed to a 24 hour PR review deadline? I haven't seen any discussion about this, but I may have missed it.

Comment (Hinton) [reply docs/contributing/pull-requests/index.md:147]: Not to my knowledge. I'm not against this though but it should be communicated.

I also think we need to spend further effort to clean up tech-leads codeowners since I haven't found a way to exclude them from review filters.

File: docs/contributing/pull-requests/branching.md
Comment (Hinton) [inline docs/contributing/pull-requests/branching.md:2]: Why are we setting sidebar positions on these? Do we dislike the alphabetical ordering?

File: docs/contributing/pull-requests/code-review.md
Comment (Hinton) [inline docs/contributing/pull-requests/code-review.md:74]: I don't think this has been established, The template repository uses a variant with emojis but it's ambiguous. I prefer the more explicit nature of conventional commits since it defines what is blocked and what isn't.

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md:147]: I agree.  I wanted to discuss here and then communicate in a bulleted list to the teams in Slack.  This felt like a good place to start the discussion.  I’m not in any way pushing for 24 - if we want to have a longer timeframe that’s fine.  My main concern is having everyone bought in on what the timeframe is so we don’t have to resort to pinging on Slack for every PR.

File: docs/contributing/pull-requests/code-review.md
Comment (Hinton) [inline docs/contributing/pull-requests/code-review.md:87]: I would argue that every review comes with this obligations. I'm worried this might associate request changes negatively.

File: docs/contributing/pull-requests/index.md
Comment (Hinton) [inline docs/contributing/pull-requests/index.md:107]: The reason this was inlined before was to get it in the same sentence. Just wanted to verify this is an intentional change.

File: docs/contributing/pull-requests/index.md
Comment (Hinton) [inline docs/contributing/pull-requests/index.md:130]: Do we want to leave a note somewhere about re-requesting review once the feedback is resolved? I've noticed it's sometimes not done which delays new reviews.

File: docs/contributing/pull-requests/index.md
Comment (Hinton) [inline docs/contributing/pull-requests/index.md:143]: I would want us to document some available integrations, and give teams a chance to start using them, before actively enforcing it.

File: docs/contributing/pull-requests/code-review.md
Comment (withinfocus) [reply docs/contributing/pull-requests/code-review.md:74]: Deliberately ambiguous too. There were a lot of comments on https://github.com/bitwarden/template/pull/5 to get it to a place folks were liking. Conventional comments (and commits) can be a lot of work in practice and if anything it's an opt-in to how you want to interact. I'd caution us to not over-prescribe here.

File: docs/contributing/pull-requests/code-review.md
Comment (withinfocus) [reply docs/contributing/pull-requests/code-review.md:87]: Can you elaborate? Not sure why this language change does anything more than expect the reviewer to stay engaged.

I too want to be sure "request changes" is taken very objectively.

File: docs/contributing/pull-requests/code-review.md
Comment (withinfocus) [inline docs/contributing/pull-requests/code-review.md:143]:
```suggestion
  - Are unit tests present?
```

File: docs/contributing/pull-requests/index.md
Comment (withinfocus) [inline docs/contributing/pull-requests/index.md:106]:
```suggestion
Once a community PR has been created, it will be automatically be linked to an internal Jira ticket. The
```

File: docs/contributing/pull-requests/index.md
Comment (withinfocus) [inline docs/contributing/pull-requests/index.md:111]:
```suggestion
When creating the PR include a Jira ticket reference so the reviewer can gain all context on the work as well as links to any associated PRs in other repositories.
```

File: docs/contributing/pull-requests/index.md
Comment (withinfocus) [reply docs/contributing/pull-requests/index.md:130]: Also support. This helps a lot, especially when many teams are involved in a review.

File: docs/contributing/pull-requests/index.md
Comment (withinfocus) [reply docs/contributing/pull-requests/index.md:143]: Handy link on more info: https://docs.github.com/en/enterprise-cloud@latest/organizations/organizing-members-into-teams/managing-code-review-settings-for-your-team

I don't know if we need anything more than what GitHub provides. Maybe Slack reminders?

File: docs/contributing/pull-requests/index.md
Comment (withinfocus) [reply docs/contributing/pull-requests/index.md:147]: Maybe "business day"? I personally feel one day is actually pretty tough unless engineers can be in formal support rotations and have this as a priority.

File: docs/contributing/pull-requests/index.md
Comment (withinfocus) [inline docs/contributing/pull-requests/index.md:152]:
```suggestion
team(s) -- or to individual engineers if assigned -- via Slack to follow up.
```

File: docs/contributing/pull-requests/index.md
Comment (withinfocus) [inline docs/contributing/pull-requests/index.md:154]: ℹ️ Other thread will potentially need updating here too.

File: docs/contributing/pull-requests/index.md
Comment (withinfocus) [inline docs/contributing/pull-requests/index.md:165]:
```suggestion
Once a Community PR has been created a Bitwarden developer will perform a code review. While we try
```

**Review by withinfocus — CHANGES_REQUESTED** [review]
Objectively requesting changes, with some subjectivity on comma usage included 😜!

File: docs/contributing/pull-requests/index.md
Comment (eliykat) [reply docs/contributing/pull-requests/index.md:147]: I'm not sure what would be a reasonable deadline here, although I agree 1 business day is tight. 2 business days?

This is not a very visible place for discussion and this would be a fairly major change to the current expectations, I suggest sharing in Slack.

File: docs/contributing/pull-requests/index.md
Comment (Hinton) [reply docs/contributing/pull-requests/index.md:143]: I get 30+ notifications daily which makes GH notifications mostly unusable without spending a non trivial amount of time maintaining it.

File: docs/contributing/pull-requests/code-review.md
Comment (Hinton) [reply docs/contributing/pull-requests/code-review.md:87]: "However, it does come with obligations for the reviewer. By blocking the PR from progressing, they have taken an ownership stake in it and should make themselves available to answer clarifying questions. They should also give clear feedback on what needs to change for the PR to get approved."

Leaving any review usually comes with this obligations. I would prefer to see this lifted into some general section. I do think we can highlight that request changes blocks the PR from moving.

File: docs/contributing/pull-requests/code-review.md
Comment (Hinton) [reply docs/contributing/pull-requests/code-review.md:74]: I've found the ambiguity to be a disservice. For example is 🎨 blocking or personal preference?

Comment (withinfocus) [reply docs/contributing/pull-requests/code-review.md:74]: A review can be a comment, approval, or changes requested, and once you push commits it resets. This is an author action that to me implies re-review, so it'd prompt that of changes requested -- blocking or not -- it's due for another look and that matter would be discussed; if it wasn't addressed by the author they consider it non-blocking. The question instead might be which side moves forward in an impasse, and that's addressed elsewhere I thought.

File: docs/contributing/pull-requests/code-review.md
Comment (trmartin4) [reply docs/contributing/pull-requests/code-review.md]: This has been fixed with the latest set of commits.

File: docs/contributing/pull-requests/code-review.md
Comment (trmartin4) [reply docs/contributing/pull-requests/code-review.md]: I tried to bridge this distinction with the changes I just pushed - I added a section at the top of the "Reviewing" section that applies to all types of review, and changed this section to indicate that "additional responsibility" is implied by blocking the PR.  Let me know how that looks.

File: docs/contributing/pull-requests/code-review.md
Comment (trmartin4) [reply docs/contributing/pull-requests/code-review.md]: Fixed in the latest commits.

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md]: Fixed in the latest commits.

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md]: I think it's back to the way it was now.  I had separated it out so that the tags were clearer, but I do see that it flows better as a single paragraph.

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md]: Fixed in the latest commits.

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md:145]: I created a section called "Addressing feedback" that includes this, as well as a recommendation to allow the commenter to "Resolve conversation" - if that's not our position I can remove that part.

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md]: Fixed in the latest commits.

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md]: Fixed in the latest commits.

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md:158]: I've added a section for this:
image.

My assumption was that the build-in reminder integration with Slack was sufficient and we didn't need to customize anything, but that may not be the case.

**Comment by trmartin4** [comment]
Thank you @Hinton, @eliykat, and @withinfocus for the feedback!

I've made some adjustments to the more trivial changes, but I feel there are a few larger outstanding items:

1. Timeframe: What is an acceptable timeframe to expect a reviewing team to perform the review before following up?  My initial off-the-cuff proposal was 24 hours, but this needs further discussion.
2. Notifications: Are the OOTB scheduled reminders from Github sufficient to give teams the visibility they need for open PRs?
3. Conventional Comments: What do we want to stay about conventional comments?  If it would be easier I can leave that section out, as it's something I assumed there was agreement on and didn't mean to divert the larger goal of this PR.

Without objections, I can raise these 3 questions to the larger group via Slack.

**Review by withinfocus — COMMENTED** [review]
Is the preview working for others? It seems to not have these changes for me. On the open topics:

1. Could we suggest "2-3 business days"? Feels like a middle ground for sometime in a business week to keep progress going.
2. I feel they are and never had anything more than a Slack integration and email where it worked for much larger organizations.
3. I suggest we leave it out. I feel there's a lot of disagreement over how much to prescribe on PR comments.

My remaining comments are super-minor.

File: docs/contributing/pull-requests/code-review.md
Comment (MGibson1) [inline docs/contributing/pull-requests/code-review.md:42]: **Suggestion:**
```suggestion
any clarifying questions or to pair on fixes suggested.
```

File: docs/contributing/pull-requests/code-review.md
Comment (MGibson1) [reply docs/contributing/pull-requests/code-review.md:80]: I prefer conventional comments to the ambiguity of emoji. It's helpful to understand the position of a commenter when addressing comments and making your position clear can save a round trip.

The issue is that a _review_ can be comment, approve, or request changes, but a comment cannot be. It's extremely rare for all of my comments to be of only one type.

Comment (MGibson1) [reply docs/contributing/pull-requests/code-review.md:80]:
```suggestion
We use [conventional comments](https://conventionalcomments.org) to help convey the intent and
expected action taken from each comment on the PR. The PR author can disagree and take different action, but clarifying your position while commenting helps make your intent clearer.
```

File: docs/contributing/pull-requests/index.md
Comment (MGibson1) [inline docs/contributing/pull-requests/index.md:162]: **issue:** This line needs to be expanded upon and supported by a section informing the author on how to create PRs for an intended change. This could be added to the branching document and linked here.

I think this is the primary reason for our large PR backlog. My recommendation would be limiting PRs to 1000 loc and preferring config or software feature flagging to allow for incomplete features to be present in the codebase.

```suggestion
  - Ask the author to split the work into a smaller PR for review. Suggest reasonable axes to on which to divide the work (see [branching for large features](#link-to-new-section)
```

**Review by MGibson1 — CHANGES_REQUESTED** [review]

File: docs/contributing/pull-requests/code-review.md
Comment (trmartin4) [reply docs/contributing/pull-requests/code-review.md]: I left the Comments section as-is, and I've broken out the changes into a separate branch to avoid having both conversations on this one.  I had assumed there was agreement on their use, so I'll use this opportunity to break apart into smaller PRs 😄 .

**Review by withinfocus — DISMISSED** [review]
Feels like this is at a good place for merge and to potentially make other improvements elsewhere.

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md]: @MGibson1, I've addressed this by completely revamping the branching page, in https://github.com/bitwarden/contributing-docs/pull/220/commits/440c84df3eaf50e502c4c2668804a8217b540b42.  As we are moving closer to a trunk-based branching scheme, with the preference toward merging to `master`, I felt that it was better to focus on how to build small, incremental changes and not detail the complex branching scenarios we had to handle previously around Short-Lived Feature Branches and Long-Lived Feature Branches.  I'd welcome any feedback on it.

I also thought that something like this might be nice to have: https://github.com/marketplace/actions/pull-request-size-labeler

Comment (MGibson1) [reply docs/contributing/pull-requests/index.md]: I think you nailed it, thank you.

I'm not really that convinced that the labels would have value, but the concept certainly does. If you think that'll help encourage people to smaller bites let's do it!

**Review by MGibson1 — DISMISSED** [review]

**Review by withinfocus — APPROVED** [review]

File: docs/contributing/pull-requests/branching.md
Comment (eliykat) [inline docs/contributing/pull-requests/branching.md:70]: (non-blocking) Are we practicing long-lived feature branches at all any more? Is there any continued justification for them? I prefer to just have one way to do things (that is, the correct way) when possible.

File: docs/contributing/pull-requests/code-review.md
Comment (eliykat) [inline docs/contributing/pull-requests/code-review.md:113]: This is a major change to our current practices, which are less strict about this. This is worth highlighting when this is shared more broadly.

File: docs/contributing/pull-requests/index.md
Comment (eliykat) [inline docs/contributing/pull-requests/index.md:138]: This is also a significant clarification (different people do different things at the moment) - would like this highlighted when these changes are shared

File: docs/contributing/pull-requests/index.md
Comment (eliykat) [reply docs/contributing/pull-requests/index.md:158]: I expect that removing tech-leads as the default reviewer should help with notification overload.

File: docs/contributing/pull-requests/index.md
Comment (eliykat) [inline docs/contributing/pull-requests/index.md:159]: I know this is going to be shared for discussion, but my 2 cents: I would prefer this be expressed in business days rather than hours. e.g. you cannot (should not) review code on public holidays or weekends.

2 business days seems fine to me as a starting point.

**Review by eliykat — DISMISSED** [review]
All comments are non-blocking, happy for you to share with the team or otherwise progress this. We can revisit these questions later if required.

**Comment by MGibson1** [comment]
@trmartin4 are you still working on this improvement?

**Comment by trmartin4** [comment]
@MGibson1 yes, actually I had this in my To Do list and it's been surpassed by some other things that have come up.

My next steps were to consolidate all the items that require highlighting as a significant change to our current process, and publish this change along with a notification in Slack.  My current list of those updates are:

- Highlight the importance of unit tests for PR approval
- The contract between the teams is that the reviewing team should act on the Github notification (through their configured notification method) and not rely on Slack posts

Am I missing anything else?

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md]: This has been addressed with [`97745a9` (#220)](https://github.com/bitwarden/contributing-docs/pull/220/commits/97745a966668222428fa4a404f7f36349ffa2d9d).

File: docs/contributing/pull-requests/branching.md
Comment (trmartin4) [reply docs/contributing/pull-requests/branching.md:70]: I feel like there is still a scenario in which large features would like to be introduced together due to either technical or product-driven considerations, but we can't put them behind a feature flag.  The Vertical Vault Nav changes are an example.  Perhaps I can supplement this with a follow-up PR that extracts this into a callout that highlights how it's not our current best practice?

**Review by withinfocus — APPROVED** [review]
My requests have long been solved.

Your `main` branch rename is probably in 500 other places ...

**Comment by trmartin4** [comment]
> Your `main` branch rename is probably in 500 other places ...

@withinfocus these should already have been fixed in `main`.  I just had to adjust the references in the part of the docs that had merge conflicts with my changes.

**Pull request merged**
